task_sdk/src/airflow/sdk/execution_time/supervisor.py

```diff
@@ -451,10 +449,38 @@ def _check_subprocess_exit(self):
 
     def _send_heartbeat_if_needed(self):
         """Send a heartbeat to the client if heartbeat interval has passed."""
-        if time.monotonic() - self._last_heartbeat >= FASTEST_HEARTBEAT_INTERVAL:
+        # If the process has exited, we don't need to send any more heartbeats
+        if self._exit_code is not None:
+            return
+
+        if time.monotonic() - self._last_heartbeat >= MIN_HEARTBEAT_INTERVAL:
             try:
                 self.client.task_instances.heartbeat(self.ti_id, pid=self._process.pid)
                 self._last_heartbeat = time.monotonic()
+            except ServerResponseError as e:
+                # TODO: Should we instead check httpx.HTTPStatusError?
+                if e.response.status_code == 409:
+                    reason = e.detail.get("reason", "")
+                    if reason == "not_running":
+                        error_msg = e.detail.get(
+                            "message", "Task is no longer in the running state and task should terminate"
+                        )
+                        log.error(error_msg, current_state=e.detail.get("current_state"))
+                    elif reason == "running_elsewhere":
+                        error_msg = e.detail.get("message", "Task is already running elsewhere")
+                        log.error(
+                            error_msg,
+                            current_hostname=e.detail.get("current_hostname"),
+                            current_pid=e.detail.get("current_pid"),
+                        )
```

Comment:
Could we simplify this and do `log.error("Server indicated we shouldn't be running anymore", detail=e.detail)`? How does structlog cope with that?

Reply:
```pycon
>>> structlog.get_logger().info("hi", detail={"reason": "not_running"})
2024-11-27 11:57:02 [info     ] hi                             detail={'reason': 'not_running'}
```

Seems okay?

Reply:
good call - resolved in https://github.com/apache/airflow/pull/44406/commits/6ed56ce1f987990b1adbc1172c2de357e45a2413